client/src/api.ts

```diff
@@ -1,12 +1,59 @@
 import useSWR, { responseInterface } from 'swr';
+import jwtDecode from 'jwt-decode';
 
-import { useStore } from './store';
+import { useStore, Action } from './store';
+
+interface DecodedToken {
+  token_type: string;
+  exp: number;
+  user_id: number;
+}
+
+/**
+ * A react hook to get a valid access token. It will refresh the token if needed.
+ * Returns the token if it's possible to get it, otherwise returns undefined.
+ */
+export async function getAccessToken(
+  tokens: { accessToken: string | null; refreshToken: string | null },
+  base: string,
+  dispatch: React.Dispatch<Action>
+): Promise<string | undefined> {
+  const { accessToken, refreshToken } = tokens;
+  // We can't get a token if we don't have a refresh token or an access token
+  if (!accessToken && !refreshToken) return undefined;
```

Comment:
I don't know how that would work in the type system. This isn't meant to enforce that we never run `getAccessToken` if we can't get one.

My intention was that making `getAccessToken` return `undefined` when an access token can't be obtained, the function explicitly communicates this state, and the API helpers will deliberately send unauthenticated requests in this case, rather than failing. This way, if a request is sent by a logged-out user, it will still be sent as long as the route/method is one for which authenticated access is allowed.

Reply:
Right. In that case, do you think throwing an error might be better API? It is best if we avoid using `undefined` to communicate anything. The helpers can catch and handle? I think that more explicitly communicates to strangers what our API is doing.

If my understanding of this is wrong, let me know.

Reply:
I don't think throwing an error would be the correct approach because being unable to generate an access token isn't always an error state - we just perform an unauthenticated request, which will succeed if we request public information. I'll change this function to return `null` to be clearer.